[masonry] Fix masonry-intrinsic-sizing-rows-001 and masonry-intrinsic-sizing-rows-006

stwrt opened the pull request:
#### 8b3d5cdb5a618ea5d7b325d2dfe73b4f9e134f60
<pre>
[masonry] Fix masonry-intrinsic-sizing-rows-001 and masonry-intrinsic-sizing-rows-006
<a href="https://bugs.webkit.org/show_bug.cgi?id=279455">https://bugs.webkit.org/show_bug.cgi?id=279455</a>
<a href="https://rdar.apple.com/problem/135736934">rdar://problem/135736934</a>

Reviewed by Sammy Gill.

There were two main issues to tackle in this PR.

1. Incorrect sizing of grids.

The grid with was not being calculated correctly. I added support for running the
track sizing algorithm when masonry is in the row direction when calculating the intrinsic logical widths.
We can place the items using their min/max-cotents to get the min/max-content of the grid.

2. Failure to stretch items inside grid.

Items were not being stretched, which appears to be due to an extra check in sizesLogicalWidthToFitContent.
We can remove this check.

* LayoutTests/TestExpectations:
* Source/WebCore/rendering/GridMasonryLayout.cpp:
 (WebCore::GridMasonryLayout::performMasonryPlacement):
 (WebCore::GridMasonryLayout::placeItemsUsingOrderModifiedDocumentOrder):
 (WebCore::GridMasonryLayout::placeItemsWithDefiniteGridAxisPosition):
 (WebCore::GridMasonryLayout::placeItemsWithIndefiniteGridAxisPosition):
 (WebCore::GridMasonryLayout::setItemGridAxisContainingBlockToGridArea):
 (WebCore::GridMasonryLayout::insertIntoGridAndLayoutItem):
* Source/WebCore/rendering/GridMasonryLayout.h:
* Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
 (WebCore::GridTrackSizingAlgorithm::gridAreaBreadthForGridItem const):
* Source/WebCore/rendering/RenderBox.cpp:
 (WebCore::RenderBox::sizesLogicalWidthToFitContent const):
* Source/WebCore/rendering/RenderGrid.cpp:
 (WebCore::RenderGrid::repeatTracksSizingIfNeeded):
 (WebCore::RenderGrid::layoutMasonry):
 (WebCore::RenderGrid::computeIntrinsicLogicalWidths const):
 (WebCore::RenderGrid::isMasonry const):
 (WebCore::RenderGrid::placeItemsOnGrid):
 (WebCore::RenderGrid::populateExplicitGridAndOrderIterator):
 (WebCore::RenderGrid::updateGridAreaForAspectRatioItems):
 (WebCore::RenderGrid::layoutGridItems):
 (WebCore::RenderGrid::layoutMasonryItems):
 (WebCore::RenderGrid::populateGridPositionsForDirection):

    Correctly sizes the last track in the masonry direction. The track was previously stuck pointing
    at the first item.

(WebCore::RenderGrid::getBaselineGridItem const):
* Source/WebCore/rendering/RenderGrid.h:

Canonical link: <a href="https://commits.webkit.org/284790@main">https://commits.webkit.org/284790@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/339fecb4790ce3d34b43de0292787daf0eff0d68

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/70536 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/49942 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/23301 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/74625 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/21714 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/72653 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/57742 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/21554 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/74625 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/21714 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/73602 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/45418 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/60810 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/74625 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/42077 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/18245 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/20075 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/64010 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/63/builds/18600 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/76345 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/14762 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/17821 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/63611 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/14805 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/60877 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/63546 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15612 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/11589 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/5234 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/45744 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/511 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/46818 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/48095 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/46560 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

sammygill commented on Source/WebCore/rendering/RenderGrid.cpp:
```diff
@@ -699,6 +699,16 @@ void RenderGrid::computeIntrinsicLogicalWidths(LayoutUnit& minLogicalWidth, Layo
+    if (isMasonry(GridTrackSizingDirection::ForColumns)) {
+        computeTrackSizesForIndefiniteSize(algorithm, GridTrackSizingDirection::ForRows, gridLayoutState, &minLogicalWidth, &maxLogicalWidth);
+
+        unsigned gridAxisTracksBeforeAutoPlacement = currentGrid().numTracks(GridTrackSizingDirection::ForRows);
```
auto

sammygill commented on Source/WebCore/rendering/RenderGrid.cpp:
```diff
@@ -699,6 +699,16 @@ void RenderGrid::computeIntrinsicLogicalWidths(LayoutUnit& minLogicalWidth, Layo
     computeTrackSizesForIndefiniteSize(algorithm, GridTrackSizingDirection::ForColumns, gridLayoutState, &minLogicalWidth, &maxLogicalWidth);
 
+    if (isMasonry(GridTrackSizingDirection::ForColumns)) {
+        computeTrackSizesForIndefiniteSize(algorithm, GridTrackSizingDirection::ForRows, gridLayoutState, &minLogicalWidth, &maxLogicalWidth);
```
The way this code is structured makes it read like we are running the track sizing algorithm twice even though that's not the case since we are only supposed to run it once according to the spec (since we would have no tracks in this direction). Can we add a comment here explaining why we need to do this in context of the masonry case below?

stwrt replied:
Sure, I can add a comment about this.

sammygill commented on Source/WebCore/rendering/RenderGrid.cpp:
```diff
@@ -699,6 +699,16 @@ void RenderGrid::computeIntrinsicLogicalWidths(LayoutUnit& minLogicalWidth, Layo
+        m_masonryLayout.performMasonryPlacement(gridAxisTracksBeforeAutoPlacement, GridTrackSizingDirection::ForColumns);
+
+        minLogicalWidth = m_masonryLayout.gridContentSize();
+        maxLogicalWidth = m_masonryLayout.gridContentSize();
```
I think we should add a comment explaining why we think it is ok to use the grid content size for both `minLogicalWidth` and `maxLogicalWidth` in this case

stwrt replied:
This has been rewritten to get min an max content based on placement.

sammygill commented on Source/WebCore/rendering/GridMasonryLayout.h:
```diff
@@ -53,11 +54,11 @@ class GridMasonryLayout {
+    void placeItemsWithIndefiniteGridAxisPosition(GridTrackSizingAlgorithm&);
+    void setItemGridAxisContainingBlockToGridArea(GridTrackSizingAlgorithm&, RenderBox&);
+    void insertIntoGridAndLayoutItem(GridTrackSizingAlgorithm&, RenderBox&, const GridArea&);
 
```
Do we expect to mutate the track sizing algorithm at all? If not should it be `const`?

sammygill commented on Source/WebCore/rendering/RenderGrid.h:
```diff
@@ -198,7 +198,7 @@ class RenderGrid final : public RenderBlock {
     void layoutMasonryItems(GridLayoutState&);
 
-    void populateGridPositionsForDirection(GridTrackSizingDirection);
+    void populateGridPositionsForDirection(GridTrackSizingAlgorithm&, GridTrackSizingDirection);
```
Should this be `const`?

sammygill commented on Source/WebCore/rendering/RenderGrid.cpp:
```diff
@@ -699,6 +699,36 @@ void RenderGrid::computeIntrinsicLogicalWidths(LayoutUnit& minLogicalWidth, Layo
+            if (grid.orderIterator().shouldSkipChild(*gridItem))
+                continue;
+
+            gridItem->setOverridingLogicalWidth(gridItem->maxPreferredLogicalWidth());
```
This code overrides the logical width of grid item's in all cases which may not be what we want? If the item has a specified logical width we probably want to keep that. I'm assuming we would only want to set the grid item's logical width to these values in certain scenarios.

stwrt replied:
As discussed, added the necessary if flags.

sammygill commented on Source/WebCore/rendering/RenderGrid.cpp:
```diff
@@ -699,6 +699,36 @@ void RenderGrid::computeIntrinsicLogicalWidths(LayoutUnit& minLogicalWidth, Layo
+        // The track sizing algorithm will only be run once in this case, since track sizing will not run in the masonry direction.
+        computeTrackSizesForIndefiniteSize(algorithm, GridTrackSizingDirection::ForRows, gridLayoutState, &minLogicalWidth, &maxLogicalWidth);
+
+        auto gridAxisTracksBeforeAutoPlacement = currentGrid().numTracks(GridTrackSizingDirection::ForRows);
```
Nit:
```suggestion
        auto gridAxisTracksCountBeforeAutoPlacement = currentGrid().numTracks(GridTrackSizingDirection::ForRows);
```

sammygill commented on Source/WebCore/rendering/RenderGrid.cpp:
```diff
@@ -699,6 +699,37 @@ void RenderGrid::computeIntrinsicLogicalWidths(LayoutUnit& minLogicalWidth, Layo
+                continue;
+
+            if (gridItem->style().logicalWidth().isAuto() || gridItem->style().logicalWidth().isPercent())
+                gridItem->setOverridingLogicalWidth(gridItem->computeReplacedLogicalWidthUsing(SizeType::MinSize, Length(LengthType::MinContent)));
```
I don't think we should be using `computedReplacedLogicalWidthUsing` here since these may not be replaced elements and I don't think that is what we are looking for here. I would expect us to use `computeIntrinsicLogicalWidthUsing` if we are looking for the min/max content sizes of the grid items.

sammygill commented on Source/WebCore/rendering/RenderGrid.cpp:
```diff
@@ -699,6 +699,37 @@ void RenderGrid::computeIntrinsicLogicalWidths(LayoutUnit& minLogicalWidth, Layo
+                continue;
+
+            if (gridItem->style().logicalWidth().isAuto() || gridItem->style().logicalWidth().isPercent())
+                gridItem->setOverridingLogicalWidth(gridItem->computeReplacedLogicalWidthUsing(SizeType::MaxSize, Length(LengthType::MaxContent)));
```
Same as above

sammygill commented on Source/WebCore/rendering/RenderGrid.cpp:
```diff
@@ -33,13 +33,14 @@
 #include "RenderTreeBuilder.h"
 #include "RenderView.h"
-#include <wtf/TZoneMallocInlines.h>
+#include "platform/Length.h"
```
We can remove this

sammygill commented on Source/WebCore/rendering/RenderBox.cpp:
```diff
@@ -2896,7 +2896,7 @@ bool RenderBox::sizesLogicalWidthToFitContent(SizeType widthType) const
 
     if (isGridItem())
-        return downcast<RenderGrid>(parent())->areMasonryColumns() || !hasStretchedLogicalWidth();
+        return (downcast<RenderGrid>(parent())->areMasonryColumns() && !GridLayoutFunctions::isOrthogonalGridItem(*(downcast<RenderGrid>(parent())), *this)) || !hasStretchedLogicalWidth();
```
I know we talked about the fact that this Masonry logic does not seem like it should belong here but we need it here temporarily to not introduce regressions, but could we please add a FIXME to indicate this?

sammygill commented on Source/WebCore/rendering/RenderBox.cpp:
```diff
@@ -2896,7 +2896,7 @@ bool RenderBox::sizesLogicalWidthToFitContent(SizeType widthType) const
 
     if (isGridItem())
-        return downcast<RenderGrid>(parent())->areMasonryColumns() || !hasStretchedLogicalWidth();
+        return (downcast<RenderGrid>(parent())->areMasonryColumns() && !GridLayoutFunctions::isOrthogonalGridItem(*(downcast<RenderGrid>(parent())), *this)) || !hasStretchedLogicalWidth();
```
Could rewrite this to improve readability if we extracted out the `downcast<RenderGrid>(parent())` to be a local here

stwrt replied:
Good idea. Fixed.

sammygill commented on Source/WebCore/rendering/RenderGrid.cpp:
```diff
@@ -33,13 +33,14 @@
 #include "GridPositionsResolver.h"
 #include "GridTrackSizingAlgorithm.h"
 #include "LayoutRepainter.h"
+#include "LayoutUnit.h"
```
Does this really not compile if we don't include this? I feel like this would have already been included by many of the other layout related headers

stwrt replied:
Removed.

sammygill commented on Source/WebCore/rendering/RenderGrid.cpp:
```diff
@@ -699,6 +700,38 @@ void RenderGrid::computeIntrinsicLogicalWidths(LayoutUnit& minLogicalWidth, Layo
+        computeTrackSizesForIndefiniteSize(algorithm, GridTrackSizingDirection::ForRows, gridLayoutState, &minLogicalWidth, &maxLogicalWidth);
+
+        auto gridAxisTracksCountBeforeAutoPlacement = currentGrid().numTracks(GridTrackSizingDirection::ForRows);
+        LayoutUnit availableLogicalWidth;
```
Since we don't actually need an `availableLogicalWidth` for the combination of arguments we are passing to `computeIntrinsicLogicalWidthUsing` can you just replace this with a `LayoutUnit()` at the call site?. 

It would be nice if that logic was extracted out into some other method that we could use directly but I don't see one readily available :/

sammygill commented on Source/WebCore/rendering/RenderGrid.cpp:
```diff
@@ -699,6 +700,38 @@ void RenderGrid::computeIntrinsicLogicalWidths(LayoutUnit& minLogicalWidth, Layo
+        auto gridAxisTracksCountBeforeAutoPlacement = currentGrid().numTracks(GridTrackSizingDirection::ForRows);
+        LayoutUnit availableLogicalWidth;
+
+        for (auto* gridItem = grid.orderIterator().first(); gridItem; gridItem = grid.orderIterator().next()) {
```
I think this whole thing deserves a comment explaining our rationale behind why we are overriding the logical width of the grid items because it is not obvious imo

sammygill commented on Source/WebCore/rendering/RenderGrid.cpp:
```diff
@@ -1559,6 +1592,9 @@ void RenderGrid::populateGridPositionsForDirection(GridTrackSizingDirection dire
             positions[i + 1] = positions[i] + offset.distributionOffset + tracks[i].unclampedBaseSize() + gap;
         positions[lastLine] = positions[nextToLastLine] + tracks[nextToLastLine].unclampedBaseSize();
 
+        if (isMasonry(direction))
```
Could you explain what this is for in the commit message? Not obvious how it's related to the changes we are making. How did this change contribute to us passing the aforementioned tests?

sammygill commented on Source/WebCore/rendering/GridMasonryLayout.h:
```diff
@@ -26,11 +26,13 @@
 
 #include "GridArea.h"
 #include "GridPositionsResolver.h"
+#include "GridTrackSizingAlgorithm.h"
```
Can we just remove this and forward declare it?

stwrt replied:
What is the benefit of the forward declaration?

sammygill commented on Source/WebCore/rendering/RenderGrid.cpp:
```diff
@@ -33,13 +33,13 @@
 #include "GridPositionsResolver.h"
 #include "GridTrackSizingAlgorithm.h"
 #include "LayoutRepainter.h"
+#include "LayoutUnit.h"
```
Is this needed?

stwrt replied:
Removed

sammygill commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
```diff
@@ -849,9 +850,13 @@ static LayoutUnit computeGridSpanSize(const Vector<GridTrack>& tracks, const Gri
+
+    if (m_renderGrid->areMasonryRows() && direction == GridTrackSizingDirection::ForRows && !GridLayoutFunctions::isOrthogonalGridItem(*m_renderGrid, gridItem))
+        return m_renderGrid->contentLogicalHeight();
+
```
While I think this is ok for now (in order to fix these tests), I think this code only makes sense if we have already computed the logical width/height of the grid. If we are currently computing the logical width of the grid (e.g. when we are computing the intrinsic widths of the grid) or before we have done masonry layout/placement, then this is probably does not return the right value. If you plan to fix this in a follow up patch then I think you should at least add a `FIXME` here.

sammygill reviewed (approved):
Please create a WPT export for the test before landing

stwrt committed:
[masonry] Fix masonry-intrinsic-sizing-rows-001 and masonry-intrinsic-sizing-rows-006
https://bugs.webkit.org/show_bug.cgi?id=279455
rdar://problem/135736934

Reviewed by Sammy Gill.

There were two main issues to tackle in this PR.

1. Incorrect sizing of grids.

The grid with was not being calculated correctly. I added support for running the
track sizing algorithm when masonry is in the row direction when calculating the intrinsic logical widths.
We can place the items using their min/max-cotents to get the min/max-content of the grid.

2. Failure to stretch items inside grid.

Items were not being stretched, which appears to be due to an extra check in sizesLogicalWidthToFitContent.
We can remove this check.

* LayoutTests/TestExpectations:
* Source/WebCore/rendering/GridMasonryLayout.cpp:
 (WebCore::GridMasonryLayout::performMasonryPlacement):
 (WebCore::GridMasonryLayout::placeItemsUsingOrderModifiedDocumentOrder):
 (WebCore::GridMasonryLayout::placeItemsWithDefiniteGridAxisPosition):
 (WebCore::GridMasonryLayout::placeItemsWithIndefiniteGridAxisPosition):
 (WebCore::GridMasonryLayout::setItemGridAxisContainingBlockToGridArea):
 (WebCore::GridMasonryLayout::insertIntoGridAndLayoutItem):
* Source/WebCore/rendering/GridMasonryLayout.h:
* Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
 (WebCore::GridTrackSizingAlgorithm::gridAreaBreadthForGridItem const):
* Source/WebCore/rendering/RenderBox.cpp:
 (WebCore::RenderBox::sizesLogicalWidthToFitContent const):
* Source/WebCore/rendering/RenderGrid.cpp:
 (WebCore::RenderGrid::repeatTracksSizingIfNeeded):
 (WebCore::RenderGrid::layoutMasonry):
 (WebCore::RenderGrid::computeIntrinsicLogicalWidths const):
 (WebCore::RenderGrid::isMasonry const):
 (WebCore::RenderGrid::placeItemsOnGrid):
 (WebCore::RenderGrid::populateExplicitGridAndOrderIterator):
 (WebCore::RenderGrid::updateGridAreaForAspectRatioItems):
 (WebCore::RenderGrid::layoutGridItems):
 (WebCore::RenderGrid::layoutMasonryItems):
 (WebCore::RenderGrid::populateGridPositionsForDirection):

    Correctly sizes the last track in the masonry direction. The track was previously stuck pointing
    at the first item.

(WebCore::RenderGrid::getBaselineGridItem const):
* Source/WebCore/rendering/RenderGrid.h:

Canonical link: https://commits.webkit.org/284790@main
